core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java

```diff
@@ -19,23 +19,665 @@
-
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.geo.GeoPoint;
+import org.elasticsearch.common.io.PathUtils;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.unit.DistanceUnit;
 import org.elasticsearch.common.xcontent.XContentParser.Token;
 import org.elasticsearch.test.ESTestCase;
+import org.hamcrest.Matcher;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.ReadableInstant;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
-import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Locale;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
 
 public abstract class BaseXContentTestCase extends ESTestCase {
 
-    public abstract XContentType xcontentType();
+    protected abstract XContentType xcontentType();
+
+    private XContentBuilder builder() throws IOException {
+        return XContentBuilder.builder(xcontentType().xContent());
+    }
+
+    public void testContentType() throws IOException {
+        assertThat(builder().contentType(), equalTo(xcontentType()));
+    }
+
+    public void testStartEndObject() throws IOException {
+        expectUnclosedException(() -> builder().startObject().bytes());
+        expectUnclosedException(() -> builder().startObject().close());
+        expectUnclosedException(() -> builder().startObject().string());
+
+        expectObjectException(() -> builder().endObject().bytes());
+        expectObjectException(() -> builder().endObject().close());
+        expectObjectException(() -> builder().endObject().string());
+
+        expectValueException(() -> builder().startObject("foo").endObject());
+        expectNonNullFieldException(() -> builder().startObject().startObject(null).endObject().endObject());
+
+        assertResult(() -> builder().startObject().endObject(), equalTo("{}"));
+        assertResult(() -> builder().startObject().startObject("foo").endObject().endObject(), equalToJSon("{'foo':{}}"));
+        assertResult(() -> builder().startObject().startObject("foo").startObject("bar").endObject().endObject().endObject(),
```

Comment:
When it gets long like this can you indent it like
```
assertResult(() -> builder()
    .startObject()
        .startObject("foo")
            .startObject("bar")
            .endObject()
        .endObject()
    .endObject(),
```

I know we `assertThat` has the matcher second, but maybe we should put the closure second for this? I think it is nice when the closure is second because it makes the code formatting prettier.